### ISSUE TITLE
EMR recreates resources when `ebs_config.volumes_per_instance` is greater than 1

### ISSUE THREAD

**Issue by software-opal** [terraform-providers/terraform-provider-aws#10446]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Terraform v0.11.13

### Affected Resource(s)

* `aws_emr_cluster`

### Terraform Configuration Files

```hcl
resource "aws_emr_cluster" "emr_cluster" {
  name          = "my-cluster"
  release_label = "emr-5.26.0"
  applications  = ["Hadoop", "Hive", "Pig", "Hue", "Spark"]

  ebs_root_volume_size = 10

  termination_protection            = false
  keep_job_flow_alive_when_no_steps = true
  scale_down_behavior               = "TERMINATE_AT_TASK_COMPLETION"

  master_instance_group {
    instance_type = "m5.xlarge"

    ebs_config {
      size                 = 32
      type                 = "gp2"
      volumes_per_instance = 2
    }
  }

  core_instance_group {
    instance_type  = "m5.xlarge"
    instance_count = 2

    ebs_config {
      size                 = 128
      type                 = "gp2"
      volumes_per_instance = 2
    }
  }
}
```

### Expected Behavior

Subsequent `terraform apply` calls should not recreate the cluster when the resource hasn't changed.

### Actual Behavior

The cluster is recreated every time a `terraform apply` is run.

```
-/+ aws_emr_cluster.emr_cluster (new resource required)
...snip...
      core_instance_group.0.ebs_config.2569541827.iops:                   "0" => "0"
      core_instance_group.0.ebs_config.2569541827.size:                   "128" => "0" (forces new resource)
      core_instance_group.0.ebs_config.2569541827.type:                   "gp2" => "" (forces new resource)
      core_instance_group.0.ebs_config.2569541827.volumes_per_instance:   "1" => "0" (forces new resource)
      core_instance_group.0.ebs_config.2986691328.iops:                   "" => ""
      core_instance_group.0.ebs_config.2986691328.size:                   "" => "128" (forces new resource)
      core_instance_group.0.ebs_config.2986691328.type:                   "" => "gp2" (forces new resource)
      core_instance_group.0.ebs_config.2986691328.volumes_per_instance:   "" => "2" (forces new resource)
      core_instance_group.0.id:                                           "ig-1MCB05JHBWE0E" => <computed>
      core_instance_group.0.instance_count:                               "2" => "2"
      core_instance_group.0.instance_type:                                "m5.xlarge" => "m5.xlarge"
      core_instance_type:                                                 "m5.xlarge" => <computed>
...snip...
      master_instance_group.0.ebs_config.2636219798.iops:                 "0" => "0"
      master_instance_group.0.ebs_config.2636219798.size:                 "32" => "0" (forces new resource)
      master_instance_group.0.ebs_config.2636219798.type:                 "gp2" => "" (forces new resource)
      master_instance_group.0.ebs_config.2636219798.volumes_per_instance: "1" => "0" (forces new resource)
      master_instance_group.0.ebs_config.3054294613.iops:                 "" => ""
      master_instance_group.0.ebs_config.3054294613.size:                 "" => "32" (forces new resource)
      master_instance_group.0.ebs_config.3054294613.type:                 "" => "gp2" (forces new resource)
      master_instance_group.0.ebs_config.3054294613.volumes_per_instance: "" => "2" (forces new resource)
      master_instance_group.0.id:                                         "ig-B0L2KTB8QGNY" => <computed>
      master_instance_group.0.instance_count:                             "1" => "1"
      master_instance_group.0.instance_type:                              "m5.xlarge" => "m5.xlarge"
...snip...
```

### Steps to Reproduce

1. `terraform apply`
2. `terraform apply`

### References

* Maybe related to #5111, #5075
* Terraform docs: https://www.terraform.io/docs/providers/aws/r/emr_cluster.html#ebs_config-3
* EMR docs for [describe-cluster](https://docs.aws.amazon.com/emr/latest/APIReference/API_DescribeCluster.html) and [list-instances](https://docs.aws.amazon.com/emr/latest/APIReference/API_ListInstances.html)

**Comment by admssa** [comment]
the same for `aws_emr_instance_group` resource. 
lifecycle ignore_changes has no effect.
PS. It's even worse:
When i try to change  this resource separately ```terraform apply -target=module.kwt-emr-dev.aws_emr_instance_group.task```

**Comment by joestump** [comment]
Just spent a little time looking. Both `aws_emr_instance_group` and `aws_emr_cluster` use the [`flattenEBSConfig`](https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_emr_cluster.go#L1731-L1751) to create/set `ebs_config`. 

Looking at the plan output I'm suspicious that `scheme.NewSet` and `resourceAwsEMRClusterEBSConfigHash` are conspiring somehow to make Terraform think the `ebs_config` set is "new" when it is clearly not:

```
core_instance_group.0.ebs_config.2569541827.iops:                   "0" => "0"
core_instance_group.0.ebs_config.2569541827.size:                   "128" => "0" (forces new resource)
core_instance_group.0.ebs_config.2569541827.type:                   "gp2" => "" (forces new resource)
core_instance_group.0.ebs_config.2569541827.volumes_per_instance:   "1" => "0" (forces new resource)
core_instance_group.0.ebs_config.2986691328.iops:                   "" => ""
core_instance_group.0.ebs_config.2986691328.size:                   "" => "128" (forces new resource)
core_instance_group.0.ebs_config.2986691328.type:                   "" => "gp2" (forces new resource)
core_instance_group.0.ebs_config.2986691328.volumes_per_instance:   "" => "2" (forces new resource)
```

**Comment by homiakos** [comment]
Is there a fix timeline or workaround?

**Comment by pmacdougall** [comment]
This is happening for me too using:
Terraform v0.12.25
provider.aws v2.63.0

**Comment by yizhu-wish** [comment]
It looks the # of volumes is hard-coded to 1?
https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_emr_cluster.go#L1745

**Comment by deedoz** [comment]
> It looks the # of volumes is hard-coded to 1?
> https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_emr_cluster.go#L1745

Agrees : 
`func flattenEBSConfig(ebsBlockDevices []*emr.EbsBlockDevice) *schema.Set {

	ebsConfig := make([]interface{}, 0)
	for _, ebs := range ebsBlockDevices {
		ebsAttrs := make(map[string]interface{})
		if ebs.VolumeSpecification.Iops != nil {
			ebsAttrs["iops"] = int(*ebs.VolumeSpecification.Iops)
		}
		if ebs.VolumeSpecification.SizeInGB != nil {
			ebsAttrs["size"] = int(*ebs.VolumeSpecification.SizeInGB)
		}
		if ebs.VolumeSpecification.VolumeType != nil {
			ebsAttrs["type"] = *ebs.VolumeSpecification.VolumeType
		}
		ebsAttrs["volumes_per_instance"] = 1

		ebsConfig = append(ebsConfig, ebsAttrs)
	}

	return schema.NewSet(resourceAwsEMRClusterEBSConfigHash, ebsConfig)
}`

ebsAttrs["volumes_per_instance"] = **1**

**Comment by c4po** [comment]
I created a PR to fix this bug and add test case to guard this feature.

**Comment by bflad** [comment]
The fix for this has been merged and will release with version 3.5.0 of the Terraform AWS Provider, later next week. Thanks to @c4po for the implementation. 👍
